### ISSUE TITLE
How can I open a new tab, window, or incognito window, next to my current session?

### ISSUE THREAD

**Issue by ifokkema** [php-webdriver/php-webdriver#796]
### What are you trying to achieve? (Expected behavior)
Open a new tab, window, or incognito window, using keyboard shortcuts.

### What do you get instead? (Actual behavior)
Nothing happens, or the letter is put into the active element.

### How could the issue be reproduced? (Steps to reproduce)
I've tried all ways I could think of to send a Control-T to the browser, but all fail. I actually need an incognito window next to my current session, but Control-T is my start to figure this out.

```php
// None of these do anything.
$driver->getKeyboard()->sendKeys(array(WebDriverKeys::CONTROL, 't'));
// Any of the following commands crash using Firefox (Facebook\WebDriver\Exception\UnknownCommandException: sendKeysToActiveElement)
$driver->getKeyboard()->pressKey(WebDriverKeys::CONTROL)
                      ->pressKey('t')
                      ->releaseKey(WebDriverKeys::CONTROL)
                      ->releaseKey('t');
$driver->action()->keyDown(null, WebDriverKeys::CONTROL)
                 ->sendKeys(null, 't')
                 ->keyUp(null, WebDriverKeys::CONTROL);

// Using an element, perhaps? Test with SHIFT.
$element = $driver->findElement(WebDriverBy::name('search_field'));
$driver->action()
       ->sendKeys($element, array(WebDriverKeys::SHIFT, 'abcde'))
       ->perform(); // Works in Chrome, although SHIFT hangs and further input is also capitals.
$driver->action()
       ->sendKeys($element, array(WebDriverKeys::CONTROL, 't'))
       ->perform(); // 'T' (note: capital).
$driver->action()
       ->sendKeys($element, array(WebDriverKeys::LEFT_CONTROL, 't'))
       ->perform(); // 'T' (note: capital).
$driver->action()
       ->keyDown($element, WebDriverKeys::CONTROL)
       ->keyDown($element, 't')
       ->keyUp($element, WebDriverKeys::CONTROL)
       ->keyUp($element, 't')
       ->perform(); // 'T' (note: capital).
```

How can I open a new tab, window, or incognito window, next to my current session? In a way that works in Chrome, and doesn't crash in Firefox?

### Details
* Php-webdriver version: 1.7.1
* PHP version: 7.2.24
* Selenium server version: 3.141.59
* Operating system: Linux/Ubuntu 18.04.4
* Browser used + version: Chromium 81.0.4044.122 or Firefox 76.0.1.

**Comment by williamdes** [comment]
This is how I did it in phpMyAdmin
Scenario:
- main tab do the work
- call dbQuery
- create or switch tabs
- do the query
- switch back

See: https://github.com/phpmyadmin/phpmyadmin/blob/d6c4c5f6d818f70af247991ee06635b49f7af1a6/test/selenium/TestBase.php#L557

Ref: https://github.com/phpmyadmin/phpmyadmin/pull/16134


```php
    /**
     * Execute a database query
     *
     * @param string       $query       SQL Query to be executed
     * @param Closure|null $onResults   The function to call when the results are displayed
     * @param Closure|null $afterSubmit The function to call after the submit button is clicked
     *
     * @return bool
     *
     * @throws Exception
     */
    public function dbQuery($query, ?Closure $onResults = null, ?Closure $afterSubmit = null): bool
    {
        $didSucceed = false;
        $handles = null;

        if (! $this->sqlWindowHandle) {
            $this->webDriver->executeScript("window.open('about:blank','_blank');", []);
            $this->webDriver->wait()->until(
                WebDriverExpectedCondition::numberOfWindowsToBe(2)
            );
            $handles = $this->webDriver->getWindowHandles();

            $lastWindow = end($handles);
            $this->webDriver->switchTo()->window($lastWindow);
            $this->login();
            $this->sqlWindowHandle = $lastWindow;
        }

        if ($handles === null) {
            $handles = $this->webDriver->getWindowHandles();
        }

        if ($this->sqlWindowHandle) {
            $this->webDriver->switchTo()->window($this->sqlWindowHandle);
            $this->byXPath('//*[contains(@class,"nav-item") and contains(., "SQL")]')->click();
            $this->waitAjax();
            $this->typeInTextArea($query);
            $this->byId('button_submit_query')->click();
            if ($afterSubmit !== null) {
                $afterSubmit->call($this);
            }
            $this->waitAjax();
            $this->waitForElement('className', 'result_query');
            // If present then
            $didSucceed = $this->isElementPresent('xpath', '//*[@class="result_query"]//*[contains(., "success")]');
            if ($onResults !== null) {
                $onResults->call($this);
            }
        }

        reset($handles);
        $lastWindow = current($handles);
        $this->webDriver->switchTo()->window($lastWindow);
        return $didSucceed;
    }
```

**Comment by williamdes** [comment]
Hiding my comment as I missed the "with keyboard shortcuts" part

**Comment by ifokkema** [comment]
If there would be a way to do it without keyboard shortcuts, I would be happy too, but JS doesn't let me open an incognito window, which is my final goal :wink: I need to make changes in the settings of my app and then check how a public viewer (hence, clean session) perceives these changes. And I prefer not to log out, log back in, log out, log back in...

**Comment by williamdes** [comment]
See: https://github.com/php-webdriver/php-webdriver/issues/793#issuecomment-630911447 to open an incognito window

I am quite sure for now you can not open the windows like you would like to do it, but I am not an expert ;)

The code I did hide will show a non keyboard method to open a new tab

**Comment by ifokkema** [comment]
Thank you, yes, I saw that one in my extensive search... but that's about starting an incognito window by default in WebDriver. So that would require me to open a whole new WebDriver session next to my existing session, which will break other things in my implementation. I'm having a strong feeling that since `SHIFT` is (sort of) working, the `CONTROL` key should work too. People have been using it (see, for instance, #226), but all implementations that I saw are very old and I fear the changes towards W3C compatibility broke the existing behavior.

**Comment by ifokkema** [comment]
Why is this issue closed? There's still no solution for it.

**Comment by andrewnicols** [comment]
@ifokkema,

The PHP Webdriver project seeks to expose the features of JSONWire (legacy) and the W3C WebDriver.

The concept of an Incognito window is browser-specific, but it can be achieved by creating an additional session.

The concept of a new Window or Tab is handled by by the New Window command (https://www.w3.org/TR/webdriver/#new-window), but this is not yet implemented in the driver. See #799 for this.

@williamdes has already provided a workaround to open new tabs for now.

**Comment by ifokkema** [comment]
Thanks for the details, @andrewnicols!

I know I've dragged two problems into one issue, which makes this a bit complicated.
- _New incognito window._ OK, if creating a new session is the only solution, I'll have to rebuild my WebDriver handling to allow two sessions in parallel.
- _Getting CTRL-key to work._ I don't mind browser-specific key combinations, I'm only running FF and Chrome anyway. SHIFT sort of works (note, sort of), CTRL not at all. I can imagine that's not top priority, but if it is something you'd like to have fixed at some point, I'd suggest keeping this issue open (or creating a new one with less clutter). For sure, it used to work (#226).

**Comment by OndraM** [comment]
@ifokkema Sorry for closing the issue previously - I thought this was solved by the previous discussion, I didn't realize it was not.

**Comment by ifokkema** [comment]
No worries, @OndraM! A lot was written and I understand you don't always have the time to carefully go through everything.

The issue is currently still closed though - I can't reopen it myself :laughing:

**Comment by ifokkema** [comment]
@OndraM would you mind re-opening this issue, or otherwise marking it as "wontfix" or so?

**Comment by andrewnicols** [comment]
The new tab/window will be fixed by #799. 

The distinction between tab and windows is a browser-specific one.

I don't know if there even is a ChromeDriver command for Incognito mode (I can't find one), but there certainly is not in either the W3C or JSONWire specifications.

Personally I don't think there's much point reopening this because the functionality will be solved by #799, and the remaining components are not possible.
